include/tvm/relay/module.h

```diff
@@ -35,32 +36,43 @@ struct Module;
  *
  *  The functional style allows users to construct custom
  *  environments easily, for example each thread can store
- *  an Module while auto-tuning.
+ *  a Module while auto-tuning.
  * */
 
 class ModuleNode : public RelayNode {
  public:
   /*! \brief A map from ids to all global functions. */
   tvm::Map<GlobalVar, Function> functions;
+  tvm::Map<GlobalTypeVar, TypeData> type_definitions;
 
   ModuleNode() {}
 
   void VisitAttrs(tvm::AttrVisitor* v) final {
     v->Visit("functions", &functions);
+    v->Visit("type_definitions", &type_definitions);
     v->Visit("global_var_map_", &global_var_map_);
+    v->Visit("global_type_var_map_", &global_type_var_map_);
   }
 
-  TVM_DLL static Module make(tvm::Map<GlobalVar, Function> global_funcs);
+  TVM_DLL static Module make(tvm::Map<GlobalVar, Function> global_funcs,
+                             tvm::Map<GlobalTypeVar, TypeData> global_type_defs);
 
   /*!
    * \brief Add a function to the global environment.
-   * \param var The name of the global function.
+   * \param var The var of the global function.
    * \param func The function.
    * \param update Controls whether you can replace a definition in the
    * environment.
    */
   void Add(const GlobalVar& var, const Function& func, bool update = false);
 
+  /*!
+   * \brief Add a type-level definition to the global environment.
+   * \param var The var of the global type definition.
+   * \param type The type definition.
+   */
+  void AddDef(const GlobalTypeVar& var, const TypeData& type);
+
```

Comment:
Not sure if I like the naming `Add`/`AddDef`. Probably better to call them something more descriptive, no? Like `AddFunc` and `AddTypeDef`, perhaps?

Reply:
Seems that the naming issue for looking up is unavoidable because it would be ambiguous whether looking up by string would be looking up a global var by name or a global type var.